Filter/set method

quantranhong1999 opened the pull request:
Compiled with no error but update rules was not succeed
Expected:
```
{
	"sessionState": "2c9f1b12-b35a-43e6-9af2-0106fb53a943",
	"methodResponses": [
		[
			"Filter/set",
			{
				"accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
				"newState": "2c9f1b12-b35a-43e6-9af2-0106fb53a943",
				"updated": {
					"singleton": {

					}
				}
			},
			"c1"
		],
		[
			"Filter/get", {
				"accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
				"list": [{
					"id": "singleton",
					"rules": [{
						"name": "My first rule",
						"condition": {
							"field": "subject",
							"comparator": "contains",
							"value": "question"
						},
						"action": {
							"appendIn": {
								"mailboxIds": ["1"]
							}
						}
					}]
				}],
				"notFound": []
			}, "c2"
		]
	]
}
```

But actual:
```
{
    "sessionState": "2c9f1b12-b35a-43e6-9af2-0106fb53a943",
    "methodResponses": [
        [
            "Filter/set",
            {
                "accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
                "newState": "2c9f1b12-b35a-43e6-9af2-0106fb53a943",
                "updated": {
                    "singleton": {
                        
                    }
                }
            },
            "c1"
        ],
        [
            "Filter/get",
            {
                "accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
                "list": [
                    {
                        "id": "singleton",
                        "rules": [
                            
                        ]
                    }
                ],
                "notFound": [
                    
                ]
            },
            "c2"
        ]
    ]
}
```

chibenwa commented on openpaas-james/jmap/src/main/scala/com/linagora/openpaas/james/jmap/method/FilterSetMethod.scala:
```diff
@@ -0,0 +1,134 @@
+
+  def update(username: Username, validatedRules: List[Rule]): SMono[FilterSetUpdateResult] = {
+    SMono.just(filteringManagement.defineRulesForUser(username, validatedRules.asJava))
+      .`then`(SMono.just(FilterSetUpdateSuccess))
```
That Why!

Try

```
SMono(filteringManagement.defineRulesForUser(username, validatedRules.asJava))
      .`then`(SMono.just(FilterSetUpdateSuccess))
```

LanKhuat commented on openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraFilterSetMethodContract.scala:
```diff
@@ -0,0 +1,123 @@
+  @Test
+  def updateShouldSucceed(): Unit = {
+    val request = s"""{
+                     |	"using": ["com:linagora:params:jmap:filter"],
+                     |	"methodCalls": [
+                     |		["Filter/set", {
+                     |			"accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
+                     |			"update": {
+                     |				"singleton": [{
+                     |					"id": "1",
+                     |					"name": "My first rule",
+                     |					"condition": {
+                     |						"field": "subject",
+                     |						"comparator": "contains",
+                     |						"value": "question"
+                     |					},
+                     |					"action": {
+                     |						"appendIn": {
+                     |							"mailboxIds": ["1"]
+                     |						}
+                     |					}
+                     |				}]
+                     |			}
+                     |		}, "c1"],
+                     |		[
+                     |			"Filter/get",
+                     |			{
+                     |				"accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
+                     |				"ids": ["singleton"]
+                     |			},
+                     |			"c2"
+                     |		]
+                     |	]
+                     |}""".stripMargin
```
Indent seems weird?

LanKhuat commented on openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraFilterSetMethodContract.scala:
```diff
@@ -0,0 +1,123 @@
+      .header(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
+      .body(request)
+    .when()
+      .post().prettyPeek()
```
remove this debug log (when you are done with the test)

quantranhong1999 commented:
Still missing validate rule(WIP)

chibenwa commented on openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraFilterSetMethodContract.scala:
```diff
@@ -120,4 +121,370 @@ trait LinagoraFilterSetMethodContract {
+      .header(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
+      .body(request)
+    .when()
+      .post().prettyPeek()
```
.prettyPeek()

chibenwa commented on openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraFilterSetMethodContract.scala:
```diff
@@ -120,4 +121,370 @@ trait LinagoraFilterSetMethodContract {
+         |	]
+         |}""".stripMargin)
+  }
+  @Test
```
Missing line break

chibenwa commented on openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraFilterSetMethodContract.scala:
```diff
@@ -120,4 +121,370 @@ trait LinagoraFilterSetMethodContract {
+      .header(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
+      .body(request)
+    .when()
+      .post().prettyPeek()
```
.prettyPeek()

chibenwa commented on openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraFilterSetMethodContract.scala:
```diff
@@ -120,4 +121,370 @@ trait LinagoraFilterSetMethodContract {
+      .header(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
+      .body(request)
+    .when()
+      .post().prettyPeek()
```
.prettyPeek()

chibenwa commented on openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraFilterSetMethodContract.scala:
```diff
@@ -120,4 +121,370 @@ trait LinagoraFilterSetMethodContract {
+         |        ]
+         |    ]
+         |}""".stripMargin)
+  }
```
Can you shoz that updating something 2 times works?

chibenwa commented on openpaas-james/jmap/src/main/scala/com/linagora/openpaas/james/jmap/method/FilterSetMethod.scala:
```diff
@@ -128,7 +126,21 @@ class FilterSetMethod @Inject()(@Named(InjectionKeys.JMAP) eventBus: EventBus,
-  def getScalaRulesFromRequest(request: FilterSetRequest): List[RuleWithId] = request.update.singleton
+//  def getScalaRulesFromRequest(request: FilterSetRequest): List[RuleWithId] = {
+//    request.parseUpdate().map()
+//  }
```
Do not commit commented code

Arsnael commented:
```
[ERROR] Failures: 
[ERROR]   DistributedLinagoraFilterSetMethodTest.createShouldFail JSON documents are different:
Different value found in node "methodResponses[0][0]", expected: <"Filter/set"> but was: <"error">.
Different keys found in node "methodResponses[0][1]", missing: "methodResponses[0][1].accountId","methodResponses[0][1].newState","methodResponses[0][1].notCreated", extra: "methodResponses[0][1].type", expected: <{"accountId":"29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6","newState":"2c9f1b12-b35a-43e6-9af2-0106fb53a943","notCreated":{"singleton":{"description":"'create' is not supported on singleton objects","type":"invalidArguments"}}}> but was: <{"type":"unknownMethod"}>

[ERROR]   DistributedLinagoraFilterSetMethodTest.destroyShouldFail JSON documents are different:
Different value found in node "methodResponses[0][0]", expected: <"Filter/set"> but was: <"error">.
Different keys found in node "methodResponses[0][1]", missing: "methodResponses[0][1].accountId","methodResponses[0][1].newState","methodResponses[0][1].notDestroyed", extra: "methodResponses[0][1].type", expected: <{"accountId":"29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6","newState":"2c9f1b12-b35a-43e6-9af2-0106fb53a943","notDestroyed":{"singleton":{"description":"'destroy' is not supported on singleton objects","type":"invalidArguments"}}}> but was: <{"type":"unknownMethod"}>

[ERROR]   DistributedLinagoraFilterSetMethodTest.filterSetShouldReturnUnknownMethodWhenOmittingCapability JSON documents are different:
Different keys found in node "methodResponses[0][1]", missing: "methodResponses[0][1].description", expected: <{"description":"Missing capability(ies): com:linagora:params:jmap:filter","type":"unknownMethod"}> but was: <{"type":"unknownMethod"}>

[ERROR]   DistributedLinagoraFilterSetMethodTest.filterSetWithWrongAccountIdShouldFail JSON documents are different:
Different value found in node "methodResponses[0][1].type", expected: <"accountNotFound"> but was: <"unknownMethod">.

[ERROR]   DistributedLinagoraFilterSetMethodTest.updateShouldFailWhenMultiplePatchObjects JSON documents are different:
Different value found in node "methodResponses[0][0]", expected: <"Filter/set"> but was: <"error">.
Different keys found in node "methodResponses[0][1]", missing: "methodResponses[0][1].accountId","methodResponses[0][1].newState","methodResponses[0][1].notUpdated","methodResponses[0][1].updated", extra: "methodResponses[0][1].type", expected: <{"accountId":"29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6","newState":"2c9f1b12-b35a-43e6-9af2-0106fb53a943","notUpdated":{"singleton2":{"description":"id singleton2 must be singleton","type":"invalidArguments"}},"updated":{"singleton":{}}}> but was: <{"type":"unknownMethod"}>

[ERROR]   DistributedLinagoraFilterSetMethodTest.updateShouldSucceed JSON documents are different:
Different value found in node "methodResponses[0][0]", expected: <"Filter/set"> but was: <"error">.
Different keys found in node "methodResponses[0][1]", missing: "methodResponses[0][1].accountId","methodResponses[0][1].newState","methodResponses[0][1].updated", extra: "methodResponses[0][1].type", expected: <{"accountId":"29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6","newState":"2c9f1b12-b35a-43e6-9af2-0106fb53a943","updated":{"singleton":{}}}> but was: <{"type":"unknownMethod"}>
Array "methodResponses[1][1].list[0].rules" has different length, expected: <1> but was: <0>.
Array "methodResponses[1][1].list[0].rules" has different content. Missing values: [{"name":"My first rule","condition":{"field":"subject","comparator":"contains","value":"question"},"action":{"appendIn":{"mailboxIds":["1"]}}}], expected: <[{"name":"My first rule","condition":{"field":"subject","comparator":"contains","value":"question"},"action":{"appendIn":{"mailboxIds":["1"]}}}]> but was: <[]>

[ERROR]   DistributedLinagoraFilterSetMethodTest.updateSomethingElseThanSingletonShouldFail JSON documents are different:
Different value found in node "methodResponses[0][0]", expected: <"Filter/set"> but was: <"error">.
Different keys found in node "methodResponses[0][1]", missing: "methodResponses[0][1].accountId","methodResponses[0][1].newState","methodResponses[0][1].notUpdated", extra: "methodResponses[0][1].type", expected: <{"accountId":"29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6","newState":"2c9f1b12-b35a-43e6-9af2-0106fb53a943","notUpdated":{"singleton2":{"description":"id singleton2 must be singleton","type":"invalidArguments"}}}> but was: <{"type":"unknownMethod"}>
```

quantranhong1999 commented:
> ```
> [ERROR] Failures: 
> [ERROR]   DistributedLinagoraFilterSetMethodTest.createShouldFail JSON documents are different:
> Different value found in node "methodResponses[0][0]", expected: <"Filter/set"> but was: <"error">.
> Different keys found in node "methodResponses[0][1]", missing: "methodResponses[0][1].accountId","methodResponses[0][1].newState","methodResponses[0][1].notCreated", extra: "methodResponses[0][1].type", expected: <{"accountId":"29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6","newState":"2c9f1b12-b35a-43e6-9af2-0106fb53a943","notCreated":{"singleton":{"description":"'create' is not supported on singleton objects","type":"invalidArguments"}}}> but was: <{"type":"unknownMethod"}>
> 
> [ERROR]   DistributedLinagoraFilterSetMethodTest.destroyShouldFail JSON documents are different:
> Different value found in node "methodResponses[0][0]", expected: <"Filter/set"> but was: <"error">.
> Different keys found in node "methodResponses[0][1]", missing: "methodResponses[0][1].accountId","methodResponses[0][1].newState","methodResponses[0][1].notDestroyed", extra: "methodResponses[0][1].type", expected: <{"accountId":"29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6","newState":"2c9f1b12-b35a-43e6-9af2-0106fb53a943","notDestroyed":{"singleton":{"description":"'destroy' is not supported on singleton objects","type":"invalidArguments"}}}> but was: <{"type":"unknownMethod"}>
> 
> [ERROR]   DistributedLinagoraFilterSetMethodTest.filterSetShouldReturnUnknownMethodWhenOmittingCapability JSON documents are different:
> Different keys found in node "methodResponses[0][1]", missing: "methodResponses[0][1].description", expected: <{"description":"Missing capability(ies): com:linagora:params:jmap:filter","type":"unknownMethod"}> but was: <{"type":"unknownMethod"}>
> 
> [ERROR]   DistributedLinagoraFilterSetMethodTest.filterSetWithWrongAccountIdShouldFail JSON documents are different:
> Different value found in node "methodResponses[0][1].type", expected: <"accountNotFound"> but was: <"unknownMethod">.
> 
> [ERROR]   DistributedLinagoraFilterSetMethodTest.updateShouldFailWhenMultiplePatchObjects JSON documents are different:
> Different value found in node "methodResponses[0][0]", expected: <"Filter/set"> but was: <"error">.
> Different keys found in node "methodResponses[0][1]", missing: "methodResponses[0][1].accountId","methodResponses[0][1].newState","methodResponses[0][1].notUpdated","methodResponses[0][1].updated", extra: "methodResponses[0][1].type", expected: <{"accountId":"29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6","newState":"2c9f1b12-b35a-43e6-9af2-0106fb53a943","notUpdated":{"singleton2":{"description":"id singleton2 must be singleton","type":"invalidArguments"}},"updated":{"singleton":{}}}> but was: <{"type":"unknownMethod"}>
> 
> [ERROR]   DistributedLinagoraFilterSetMethodTest.updateShouldSucceed JSON documents are different:
> Different value found in node "methodResponses[0][0]", expected: <"Filter/set"> but was: <"error">.
> Different keys found in node "methodResponses[0][1]", missing: "methodResponses[0][1].accountId","methodResponses[0][1].newState","methodResponses[0][1].updated", extra: "methodResponses[0][1].type", expected: <{"accountId":"29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6","newState":"2c9f1b12-b35a-43e6-9af2-0106fb53a943","updated":{"singleton":{}}}> but was: <{"type":"unknownMethod"}>
> Array "methodResponses[1][1].list[0].rules" has different length, expected: <1> but was: <0>.
> Array "methodResponses[1][1].list[0].rules" has different content. Missing values: [{"name":"My first rule","condition":{"field":"subject","comparator":"contains","value":"question"},"action":{"appendIn":{"mailboxIds":["1"]}}}], expected: <[{"name":"My first rule","condition":{"field":"subject","comparator":"contains","value":"question"},"action":{"appendIn":{"mailboxIds":["1"]}}}]> but was: <[]>
> 
> [ERROR]   DistributedLinagoraFilterSetMethodTest.updateSomethingElseThanSingletonShouldFail JSON documents are different:
> Different value found in node "methodResponses[0][0]", expected: <"Filter/set"> but was: <"error">.
> Different keys found in node "methodResponses[0][1]", missing: "methodResponses[0][1].accountId","methodResponses[0][1].newState","methodResponses[0][1].notUpdated", extra: "methodResponses[0][1].type", expected: <{"accountId":"29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6","newState":"2c9f1b12-b35a-43e6-9af2-0106fb53a943","notUpdated":{"singleton2":{"description":"id singleton2 must be singleton","type":"invalidArguments"}}}> but was: <{"type":"unknownMethod"}>
> ```

yeh the tests are not done yet

quantranhong1999 commented:
Have a error when combine Filter/set and filter/get (only happen for distributed server):
```
{
    "sessionState": "2c9f1b12-b35a-43e6-9af2-0106fb53a943",
    "methodResponses": [
        [
            "Filter/set",
            {
                "accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
                "newState": "2c9f1b12-b35a-43e6-9af2-0106fb53a943",
                "updated": {
                    "singleton": {
                        
                    }
                }
            },
            "c1"
        ],
        [
            "error",
            {
                "type": "invalidArguments",
                "description": "UUID string too large"
            },
            "c2"
        ]
    ]
}
```
Although the mailboxId which I used is normal, maybe something like out of memory when parsing UUID?

chibenwa commented:
Please do some debugging to further understand the issue...

quantranhong1999 commented:
> Have a error when combine Filter/set and filter/get (only happen for distributed server):
> 
> ```
> {
>     "sessionState": "2c9f1b12-b35a-43e6-9af2-0106fb53a943",
>     "methodResponses": [
>         [
>             "Filter/set",
>             {
>                 "accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
>                 "newState": "2c9f1b12-b35a-43e6-9af2-0106fb53a943",
>                 "updated": {
>                     "singleton": {
>                         
>                     }
>                 }
>             },
>             "c1"
>         ],
>         [
>             "error",
>             {
>                 "type": "invalidArguments",
>                 "description": "UUID string too large"
>             },
>             "c2"
>         ]
>     ]
> }
> ```
> 
> Will working on this error

#https://github.com/linagora/openpaas-james/pull/67/commits/f8558183b9b8cdb0bb34e5eec774f3a7aa8d45d9 fixed this :D

chibenwa commented on openpaas-james/jmap/src/main/scala/com/linagora/openpaas/james/jmap/model/FilterGet.scala:
```diff
@@ -45,5 +45,5 @@ object Rule {
   def fromMailboxIds(mailboxIds: ImmutableList[String], mailboxIdFactory: MailboxId.Factory) : List[MailboxId] = List(mailboxIdFactory.fromString(mailboxIds.get(0)))
   def convertListMailboxIdToListString(mailboxIds: List[MailboxId]) : List[String] =
-    mailboxIds.collect[String](id => String.valueOf(id))
+    mailboxIds.map(mailboxId => mailboxId.serialize())
```
`mailboxIds.map(_.serialie)`?

chibenwa commented on openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraFilterSetMethodContract.scala:
```diff
@@ -485,4 +485,327 @@ trait LinagoraFilterSetMethodContract {
+      .header(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
+      .body(request)
+    .when()
+      .post().prettyPeek()
```
debug spotted

chibenwa commented on openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraFilterSetMethodContract.scala:
```diff
@@ -485,4 +485,327 @@ trait LinagoraFilterSetMethodContract {
+      .header(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
+      .body(request)
+    .when()
+      .post().prettyPeek()
```
oups

chibenwa commented on openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraFilterSetMethodContract.scala:
```diff
@@ -485,4 +485,327 @@ trait LinagoraFilterSetMethodContract {
+      .header(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
+      .body(request)
+    .when()
+      .post().prettyPeek()
```
oups

chibenwa commented on openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraFilterSetMethodContract.scala:
```diff
@@ -485,4 +485,327 @@ trait LinagoraFilterSetMethodContract {
+      .header(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
+      .body(request)
+    .when()
+      .post().prettyPeek()
```
oups

chibenwa commented on openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraFilterSetMethodContract.scala:
```diff
@@ -485,4 +485,327 @@ trait LinagoraFilterSetMethodContract {
+      .header(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
+      .body(request)
+    .when()
+      .post().prettyPeek()
```
oups

chibenwa commented on openpaas-james/jmap/src/main/scala/com/linagora/openpaas/james/jmap/method/FilterSetMethod.scala:
```diff
@@ -105,7 +105,17 @@ class FilterSetMethod @Inject()(val metricFactory: MetricFactory,
     SFlux.fromIterable(filterSetRequest.parseUpdate()
       .map[SMono[FilterSetUpdateResult]]({
-        case (id, Right(update)) => updateRules(mailboxSession.getUser, RuleWithId.toJava(update.rules))
+        case (id, Right(update)) => {
```
A more idiomtic way would be

```
   SFlux.fromIterable(filterSetRequest.parseUpdate()
      .map[SMono[FilterSetUpdateResult]] {
            ensureNoDuplicatedRules(update.rules)
            ensureNoMultipleMailboxesRules(update.rules)
            updateRules(mailboxSession.getUser, RuleWithId.toJava(update.rules))
       }
       .onErrorResume(e => SMono.just(FilterSetUpdateFailure(id, e))
```

chibenwa commented on openpaas-james/jmap/src/main/scala/com/linagora/openpaas/james/jmap/method/FilterSetMethod.scala:
```diff
@@ -128,4 +138,14 @@ class FilterSetMethod @Inject()(val metricFactory: MetricFactory,
         .toMap)
+
+  def ensureNoDuplicatedRules(rules: List[RuleWithId]) =
+    if (!rules.distinctBy(_.id).length.equals(rules.length)) throw new DuplicatedRuleException("There are some duplicated rules")
```
Use blocks `{}`

chibenwa commented on openpaas-james/jmap/src/main/scala/com/linagora/openpaas/james/jmap/method/FilterSetMethod.scala:
```diff
@@ -128,4 +138,14 @@ class FilterSetMethod @Inject()(val metricFactory: MetricFactory,
         Some(SetErrorDescription("'destroy' is not supported on singleton objects")))))
         .toMap)
+
+  def ensureNoDuplicatedRules(rules: List[RuleWithId]) =
```
We need to use Either here to get the validation down to the type system

chibenwa commented on openpaas-james/jmap/src/main/scala/com/linagora/openpaas/james/jmap/method/FilterSetMethod.scala:
```diff
@@ -128,4 +138,14 @@ class FilterSetMethod @Inject()(val metricFactory: MetricFactory,
+  def ensureNoDuplicatedRules(rules: List[RuleWithId]) =
+    if (!rules.distinctBy(_.id).length.equals(rules.length)) throw new DuplicatedRuleException("There are some duplicated rules")
+
+  def ensureNoMultipleMailboxesRules(rules: List[RuleWithId]) = {
```
We need to use Either here to get the validation down to the type system

quantranhong1999 commented:
Can someone check this? The tests were supposed to be green

Arsnael commented:
```
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.011 s <<< FAILURE! - in com.linagora.openpaas.james.DistributedLinagoraFilterSetMethodTest
[ERROR] com.linagora.openpaas.james.DistributedLinagoraFilterSetMethodTest  Time elapsed: 0.011 s  <<< ERROR!
java.lang.NoClassDefFoundError: org/apache/james/backends/es/v7/DockerElasticSearchSingleton
	at com.linagora.openpaas.james.DistributedLinagoraFilterSetMethodTest.<clinit>(DistributedLinagoraFilterSetMethodTest.java:40)
Caused by: java.lang.ClassNotFoundException: org.apache.james.backends.es.v7.DockerElasticSearchSingleton
	at com.linagora.openpaas.james.DistributedLinagoraFilterSetMethodTest.<clinit>(DistributedLinagoraFilterSetMethodTest.java:40)
```
I thought we were still using the backport of es6 for now, so I'm not too sure what is going on here? Would need to dig deeper

Arsnael commented on openpaas-james/integration-tests/jmap/distributed-jmap-integration-tests/src/test/java/com/linagora/openpaas/james/DistributedLinagoraFilterSetMethodTest.java:
```diff
@@ -0,0 +1,67 @@
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
```
why apache license here?

quantranhong1999 replied:
removed

Arsnael commented:
@quantranhong1999 you mixed up deps here a bit. You were using in your test the docker ES from apache james (based on ES7 now) while we are still on the backport for now in tmail. Need to use the docker ES that Lan duplicated in our code, based on the ES6 backport.

Don't hesitate as well to do a `git submodule update` after rebasing to be sure you run on the apache james version used by the branch, that I just updated :)

chibenwa commented on openpaas-james/integration-tests/jmap/distributed-jmap-integration-tests/src/test/java/com/linagora/openpaas/james/DistributedLinagoraFilterSetMethodTest.java:
```diff
@@ -14,6 +13,7 @@
 
 import com.linagora.openpaas.james.app.DistributedJamesConfiguration;
 import com.linagora.openpaas.james.app.DistributedServer;
+import com.linagora.openpaas.james.app.DockerElasticSearchExtension;
```
Knew it!

Arsnael commented:
Aaaaaaand... it's green!
